internal/clientconn/listener.go

```diff
@@ -69,23 +74,71 @@ func NewListener(opts *NewListenerOpts) *Listener {
 func (l *Listener) Run(ctx context.Context) error {
 	logger := l.opts.Logger.Named("listener")
 
-	var err error
-	if l.listener, err = net.Listen("tcp", l.opts.ListenAddr); err != nil {
-		return lazyerrors.Error(err)
+	useSock := l.opts.ListenSock != ""
+	useTcp := !useSock || l.opts.ListenAddr != defaultListeningIP
+
+	if useTcp {
+		var err error
+		if l.tcpListener, err = net.Listen("tcp", l.opts.ListenAddr); err != nil {
+			return lazyerrors.Error(err)
+		}
+	}
+
+	if useSock {
+		var err error
+		if l.sockListener, err = net.Listen("unix", l.opts.ListenSock); err != nil {
+			return lazyerrors.Error(err)
+		}
 	}
 
 	close(l.listening)
-	logger.Sugar().Infof("Listening on %s ...", l.Addr())
+
+	if useTcp {
+		logger.Sugar().Infof("Listening on %s ...", l.Addr())
+	}
+
+	if useSock {
+		logger.Sugar().Infof("Listening on %s ...", l.Sock())
+	}
 
 	// handle ctx cancellation
 	go func() {
 		<-ctx.Done()
-		l.listener.Close()
+
+		if useTcp {
+			l.tcpListener.Close()
+		}
+
+		if useSock {
+			l.sockListener.Close()
+		}
 	}()
 
 	var wg sync.WaitGroup
+
+	// handle TCP stream
+	if useTcp {
+		wg.Add(1)
+		go listenLoop(ctx, &wg, l, l.tcpListener, logger)
```

Comment:
In this particular place, it's not a big deal, but we pass `wg` through two functions (`listenLoop` and then `runConn`) to just call `wg.Done` somewhere at the end. Would it be possible to refactor it to something like 

```
go func() {
  listenLoop(ctx, l, l.tcpListener, logger)
  wg.Done()
}
```

and then we don't need to pass `wg` through multiple functions?

Reply:
Yep also was think about it.
But left it.

Reply:
Because we will need to make this 2 times, which is also.... a bit of duplication.

Reply:
@zhiburt, no worries :)